pymc_marketing/mmm/delayed_saturated_mmm.py

```diff
@@ -299,33 +280,135 @@ def build_model(
+                        continue
+
+                    length = dimensions.get(config.get("dims", [None, None])[1], 1)
+                    priors[param] = self.create_tvp_priors(param, config, length, positive=is_positive)
+                    continue
+
+                dist_func = getattr(pm, prior_type, None)
+                if not dist_func: raise ValueError(f"Invalid distribution type {prior_type}")
+                config_copy = {k: v for k, v in config.items() if k != "type"}
+                priors[param] = dist_func(name=param, **config_copy)
+
+        return priors
+
+    def create_likelihood(self, model_config, target_, mu):
+        likelihood_config = model_config.get("likelihood", {})
+        likelihood_type = likelihood_config.get("type")
+        dims = likelihood_config.get("dims")
+
+        if not likelihood_type:
+            raise ValueError("Likelihood type must be specified in the model config.")
+
+        likelihood_func = getattr(pm, likelihood_type, None)
+        if likelihood_func is None:
+            raise ValueError(f"Invalid likelihood type {likelihood_type}")
+
+        # Transform mu if the likelihood type is Lognormal or HurdleLognormal
+        if likelihood_type in ['LogNormal', 'HurdleLogNormal']:
+            mu = pt.log(mu)
+
+        # Create sub-priors
+        sub_priors = {}
+        for param, config in likelihood_config.items():
+            if param not in ['type', 'dims']:  # Skip 'type' and 'dims'
+                if param == 'params':  # Handle nested 'params'
+                    for sub_param, sub_config in config.items():
+                        sub_priors[sub_param] = self.create_priors_from_config({sub_param: sub_config})[sub_param]
+                else:
+                    sub_priors[param] = self.create_priors_from_config({param: config})[param]
+
+        return likelihood_func(name="likelihood", mu=mu, observed=target_, dims=dims,  **sub_priors)
+
+    def create_tvp_priors(self, param, config, length, positive=False):
+        dims = config.get("dims", None)  # Extracting dims from the config
+        print(dims)
+        gp_list = [self.gp_wrapper(name=f"{param}_{i}", X=np.arange(len(self.X[self.date_column]))[:, None], config=config, positive=positive) for i in range(length)]
+        stacked_gp = pt.stack(gp_list, axis=1)
+        return pm.Deterministic(f"{param}", stacked_gp, dims=dims)
+
+
+    def gp_wrapper(self, name, X, config, positive=False, **kwargs):
+        return self.gp_coeff(X, name, config=config, positive=positive, **kwargs)
+
+    def gp_coeff(self, X, name, mean=0.0, positive=False, config=None):
+        params = pm.find_constrained_prior(pm.Gamma, 8, 12, init_guess={"alpha": 1, "beta": 1}, mass=0.8)
+        ell = pm.Gamma(f"ell_{name}", **params)
+        eta = pm.Exponential(f"_eta_{name}", lam=1)
+  #      cov = eta ** 2 * pm.gp.cov.ExpQuad(1, ls=ell)
+
+        cov = eta ** 2 * pm.gp.cov.Matern32(1, ls=ell)
```

Comment:
I don't get this. `eta ** 2` is going to blow up for negative x. Plotting it for x' = 0 and lam = 1:
<img width="606" alt="Screenshot 2024-01-16 at 16 50 51" src="https://github.com/pymc-labs/pymc-marketing/assets/111059273/3d6037dd-9212-4c9e-9f84-c8c07a2f6609">

Is the point to only let coef value at t = 0 be influenced by coef values at t < 0?

Reply:
Hey, apologies, I've been a bit slow on this one. There's been quite a few changes to the main with the custom priors/likelihood pr & shortly more with the out-of-sample prediction. 

@bwengals is going to be helping out implementing this one too

Reply:
No worries. For context, I'm here because I'm adding TVC to my company's MMM, and thought it would make sense to drop review comments for clarification sake (and to give back, thanks for the beautiful lib). But if any of these comments are out of place, please just ignore them.